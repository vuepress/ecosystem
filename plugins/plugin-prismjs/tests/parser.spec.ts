import { describe, expect, it } from 'vitest'
import {
  highlightLines,
  notationDiff,
  notationErrorLevel,
  notationFocus,
  notationHighlight,
  parse,
  resolveHighlightLines,
} from '../src/node/index.js'

const genCode = (code: string) => `<pre><code>${code}</code></pre>`

describe('@vuepress/plugin-prismjs > parser', () => {
  it('normal parse', () => {
    const code = genCode(`const a = 1
const b = 2
const c = 3
`)
    const parser = parse(code)
    expect(parser.lines.length).toBe(4)
    expect(parser.pre.before).toContain('<pre')
    expect(parser.lines.every((line) => line.class.includes('line'))).toBe(true)

    expect(parser.stringify()).toMatchSnapshot()
  })

  it('normal parse line node add class', () => {
    const code = genCode(`  const a = 1
  const b = 2
  console.log(a + b)`)
    const parser = parse(code)

    parser.line((node) => node.class.push('highlighted'))

    expect(parser.stringify()).toMatchSnapshot()
  })

  it('parse highlight lines', () => {
    const code = genCode(`const a = 1
const b = 2
const c = 3
console.log(a + b)

function add(a, b) {
  return a + b
}
`)
    const parser = parse(code)

    highlightLines(parser, resolveHighlightLines('{1,3,6-8}'))

    const result = parser.stringify()

    expect(parser.lines[0].class.includes('highlighted')).toBe(true) // line 1
    expect(parser.lines[6].class.includes('highlighted')).toBe(true) // line 7
    expect(parser.lines[4].class.includes('highlighted')).toBe(false) // line 4

    expect(result).toMatchSnapshot()
  })

  it('parse notation highlight', () => {
    const code = genCode(`const a = 1
const b = 2
const c = 3 // [!code highlight]
console.log(a + b) // [!code hl]

function add(a, b) {
  return a + b
}
`)
    const parser = parse(code)
    notationHighlight(parser)

    const result = parser.stringify()

    expect(parser.lines[2].class.includes('highlighted')).toBe(true) // line 3
    // magic comment should be removed
    expect(parser.lines[2].content.includes('// [!code highlight]')).toBe(false) // line 3
    expect(parser.lines[3].class.includes('highlighted')).toBe(true) // line 4

    // pre tag should add has-highlighted class
    expect(parser.pre.class.includes('has-highlighted')).toBe(true)

    expect(result).toMatchSnapshot()
  })

  it('parse notation highlight:number', () => {
    const code = genCode(`const a = 1
const b = 2 // [!code highlight:3]
const c = 3
console.log(a + b)

function add(a, b) {
  return a + b
}
`)
    const parser = parse(code)
    notationHighlight(parser)

    const result = parser.stringify()

    expect(
      parser.lines
        .slice(1, 4)
        .every((line) => line.class.includes('highlighted')),
    ).toBe(true)

    expect(parser.lines[4].class.includes('highlighted')).toBe(false)

    expect(result).toMatchSnapshot()
  })

  it('parse notation diff', () => {
    const code = genCode(`const a = 1
const b = 2 // [!code ++]
const c = 3
console.log(a + b) // [!code --]

function add(a, b) { // [!code ++]
  return a + b
}
`)
    const parser = parse(code)
    notationDiff(parser)

    const result = parser.stringify()

    expect(parser.lines[1].class.includes('diff add')).toBe(true) // line 2
    expect(parser.lines[3].class.includes('diff remove')).toBe(true) // line 4

    expect(parser.lines[5].content.includes('// [!code ++]')).toBe(false)

    expect(parser.pre.class.includes('has-diff')).toBe(true)

    expect(result).toMatchSnapshot()
  })

  it('parse notation diff ++:number', () => {
    const code = genCode(`const a = 1
const b = 2 // [!code ++:3]
const c = 3
console.log(a + b)

function add(a, b) {
  return a + b
}
`)
    const parser = parse(code)
    notationDiff(parser)

    const result = parser.stringify()

    expect(
      parser.lines.slice(1, 4).every((line) => line.class.includes('diff add')),
    ).toBe(true)

    expect(parser.lines[4].class.includes('diff add')).toBe(false)

    expect(result).toMatchSnapshot()
  })

  it('parse notation focus', () => {
    const code = genCode(`const a = 1
const b = 2 // [!code focus]
const c = 3
console.log(a + b)

function add(a, b) { // [!code focus]
  return a + b
}
`)
    const parser = parse(code)
    notationFocus(parser)

    const result = parser.stringify()

    expect(parser.lines[1].class.includes('focused')).toBe(true) // line 2
    expect(parser.lines[5].class.includes('focused')).toBe(true) // line 5

    expect(parser.pre.class.includes('has-focused')).toBe(true)

    expect(result).toMatchSnapshot()
  })

  it('parse notation focus:number', () => {
    const code = genCode(`const a = 1
const b = 2 // [!code focus:3]
const c = 3
console.log(a + b)

function add(a, b) {
  return a + b
}
`)
    const parser = parse(code)
    notationFocus(parser)

    const result = parser.stringify()

    expect(
      parser.lines.slice(1, 4).every((line) => line.class.includes('focused')),
    ).toBe(true)

    expect(parser.lines[4].class.includes('focused')).toBe(false)

    expect(result).toMatchSnapshot()
  })

  it('parse notation error level', () => {
    const code = genCode(`const a = 1
const b = 2 // [!code warning]
const c = 3
console.log(a + b) // [!code error]

function add(a, b) { // [!code warning]
  return a + b
}
`)
    const parser = parse(code)
    notationErrorLevel(parser)

    const result = parser.stringify()

    expect(parser.lines[1].class.includes('warning')).toBe(true) // line 2
    expect(parser.lines[3].class.includes('error')).toBe(true) // line 4

    expect(parser.lines[5].content.includes('// [!code warning]')).toBe(false)

    expect(parser.pre.class.includes('has-highlighted')).toBe(true)

    expect(result).toMatchSnapshot()
  })

  it('parse notation diff error:number', () => {
    const code = genCode(`const a = 1
const b = 2 // [!code error:3]
const c = 3
console.log(a + b)

function add(a, b) {
  return a + b
}
`)
    const parser = parse(code)
    notationErrorLevel(parser)

    const result = parser.stringify()

    expect(
      parser.lines.slice(1, 4).every((line) => line.class.includes('error')),
    ).toBe(true)

    expect(parser.lines[4].class.includes('error')).toBe(false)

    expect(result).toMatchSnapshot()
  })
})
