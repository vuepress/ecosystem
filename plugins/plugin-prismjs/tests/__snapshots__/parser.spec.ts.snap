// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`@vuepress/plugin-prismjs > parser > normal parse 1`] = `
"<pre><code><span class="line">const a = 1</span>
<span class="line">const b = 2</span>
<span class="line">const c = 3</span>
<span class="line"></code></pre></span>"
`;

exports[`@vuepress/plugin-prismjs > parser > normal parse line node add class 1`] = `
"<pre><code><span class="line highlighted">  const a = 1</span>
<span class="line highlighted">  const b = 2</span>
<span class="line highlighted">  console.log(a + b)</code></pre></span>"
`;

exports[`@vuepress/plugin-prismjs > parser > parse highlight lines 1`] = `
"<pre><code><span class="line highlighted">const a = 1</span>
<span class="line">const b = 2</span>
<span class="line highlighted">const c = 3</span>
<span class="line">console.log(a + b)</span>
<span class="line"></span>
<span class="line highlighted">function add(a, b) {</span>
<span class="line highlighted">  return a + b</span>
<span class="line highlighted">}</span>
<span class="line"></code></pre></span>"
`;

exports[`@vuepress/plugin-prismjs > parser > parse notation diff ++:number 1`] = `
"<pre class="has-diff"><code><span class="line">const a = 1</span>
<span class="line diff add">const b = 2</span>
<span class="line diff add">const c = 3</span>
<span class="line diff add">console.log(a + b)</span>
<span class="line"></span>
<span class="line">function add(a, b) {</span>
<span class="line">  return a + b</span>
<span class="line">}</span>
<span class="line"></code></pre></span>"
`;

exports[`@vuepress/plugin-prismjs > parser > parse notation diff 1`] = `
"<pre class="has-diff"><code><span class="line">const a = 1</span>
<span class="line diff add">const b = 2</span>
<span class="line">const c = 3</span>
<span class="line diff remove">console.log(a + b)</span>
<span class="line"></span>
<span class="line diff add">function add(a, b) {</span>
<span class="line">  return a + b</span>
<span class="line">}</span>
<span class="line"></code></pre></span>"
`;

exports[`@vuepress/plugin-prismjs > parser > parse notation diff error:number 1`] = `
"<pre class="has-highlighted"><code><span class="line">const a = 1</span>
<span class="line highlighted error">const b = 2</span>
<span class="line highlighted error">const c = 3</span>
<span class="line highlighted error">console.log(a + b)</span>
<span class="line"></span>
<span class="line">function add(a, b) {</span>
<span class="line">  return a + b</span>
<span class="line">}</span>
<span class="line"></code></pre></span>"
`;

exports[`@vuepress/plugin-prismjs > parser > parse notation error level 1`] = `
"<pre class="has-highlighted"><code><span class="line">const a = 1</span>
<span class="line highlighted warning">const b = 2</span>
<span class="line">const c = 3</span>
<span class="line highlighted error">console.log(a + b)</span>
<span class="line"></span>
<span class="line highlighted warning">function add(a, b) {</span>
<span class="line">  return a + b</span>
<span class="line">}</span>
<span class="line"></code></pre></span>"
`;

exports[`@vuepress/plugin-prismjs > parser > parse notation focus 1`] = `
"<pre class="has-focused-lines"><code><span class="line">const a = 1</span>
<span class="line has-focus">const b = 2</span>
<span class="line">const c = 3</span>
<span class="line">console.log(a + b)</span>
<span class="line"></span>
<span class="line has-focus">function add(a, b) {</span>
<span class="line">  return a + b</span>
<span class="line">}</span>
<span class="line"></code></pre></span>"
`;

exports[`@vuepress/plugin-prismjs > parser > parse notation focus:number 1`] = `
"<pre class="has-focused-lines"><code><span class="line">const a = 1</span>
<span class="line has-focus">const b = 2</span>
<span class="line has-focus">const c = 3</span>
<span class="line has-focus">console.log(a + b)</span>
<span class="line"></span>
<span class="line">function add(a, b) {</span>
<span class="line">  return a + b</span>
<span class="line">}</span>
<span class="line"></code></pre></span>"
`;

exports[`@vuepress/plugin-prismjs > parser > parse notation highlight 1`] = `
"<pre class="has-highlighted"><code><span class="line">const a = 1</span>
<span class="line">const b = 2</span>
<span class="line highlighted">const c = 3</span>
<span class="line highlighted">console.log(a + b)</span>
<span class="line"></span>
<span class="line">function add(a, b) {</span>
<span class="line">  return a + b</span>
<span class="line">}</span>
<span class="line"></code></pre></span>"
`;

exports[`@vuepress/plugin-prismjs > parser > parse notation highlight:number 1`] = `
"<pre class="has-highlighted"><code><span class="line">const a = 1</span>
<span class="line highlighted">const b = 2</span>
<span class="line highlighted">const c = 3</span>
<span class="line highlighted">console.log(a + b)</span>
<span class="line"></span>
<span class="line">function add(a, b) {</span>
<span class="line">  return a + b</span>
<span class="line">}</span>
<span class="line"></code></pre></span>"
`;

exports[`@vuepress/plugin-prismjs > parser > should not remove // [\\!code xxx] 1`] = `"<pre><code><span class="line">const a = 1 // [!code focus]</code></pre></span>"`;

exports[`@vuepress/plugin-prismjs > parser > should remove single line // [!code xxx] 1`] = `
"<pre class="has-focused-lines"><code><span class="line">const a = 1</span>
<span class="line has-focus">const b = 2</code></pre></span>"
`;
